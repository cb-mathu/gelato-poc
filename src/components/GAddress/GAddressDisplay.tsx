import { defineComponent } from 'vue';
import { ADDRESS_PROPS } from './constants';
import { Address, Country, MergedAddressProp, State } from './types';
import { getCustomerName, removeLastComma } from './utils';
import { Clickable } from '../../utils/accessibility';
import { isEmpty } from '../../utils/common';

const GAddressDisplay = defineComponent({
  name: 'g-address-display',
  props: ADDRESS_PROPS,
  emits: ['edit'],
  setup(props, { emit }) {
    return () => {
      const address = props.address as Address;
      const getAddressLine = (...fields: string[]) => {
        const fieldValues = fields.map((field: string, index: number) => {
          let addressField: string = props.address?.[field];
          if (!(typeof props.address?.[field] === 'string')) {
            switch (field) {
              case 'country':
                addressField = (props.address?.[field] as Country)?.country_name;
                break;
              case 'state':
                addressField = (props.address?.[field] as State)?.state_name;
                break;
            }
          }
          return removeLastComma(addressField);
        });
        return fieldValues.map((fieldValue: string | undefined, index: number) => {
          return fieldValue && <span class={`g-address_${fields[index]} g-address__value`}>{`${fieldValue}`}</span>;
        });
      };
      return (
        <div class="g-address">
          <div class="g-address__title">
            <h3>{getCustomerName(address)}</h3>
            {props.showEdit && (
              <Clickable as="a" onClick={() => emit('edit')}>
                Edit
              </Clickable>
            )}
          </div>
          <h6 class="g-address__caption">{address.company}</h6>
          <p class="g-address__content">
            <p>{getAddressLine('line1', 'line2', 'line3')}</p>

            <p>{getAddressLine('city', 'zip')}</p>

            <p>{getAddressLine('state', 'country')}</p>
          </p>
        </div>
      );
    };
  },
});

export default GAddressDisplay as {
  new (): {
    $props: MergedAddressProp;
  };
};
