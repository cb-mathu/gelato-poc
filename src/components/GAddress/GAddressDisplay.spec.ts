import { render, screen, fireEvent } from '@testing-library/vue';
import * as stories from './GAddressDisplay.stories';

const { DefaultAddress } = stories;

describe('GAddressDisplay : ', () => {
  test('renders default address with all values', () => {
    const { container } = render(DefaultAddress());
    expect(container.firstChild).toMatchSnapshot();
  });

  test('calls onEdit prop when clicked', () => {
    const handleClick = jest.fn();
    render(DefaultAddress({ onEdit: handleClick }));
    fireEvent.click(screen.getByRole('button'));
    expect(handleClick).toHaveBeenCalled();
  });
});
