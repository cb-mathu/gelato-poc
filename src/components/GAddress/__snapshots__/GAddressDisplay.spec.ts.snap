// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GAddressDisplay :  renders default address with all values 1`] = `
<div
  class="g-address"
>
  <div
    class="g-address__title"
  >
    <h3>
      John Doe
    </h3>
    <a
      as="a"
      role="button"
      tabindex="0"
    >
      Edit
    </a>
  </div>
  <h6
    class="g-address__caption"
  >
    Chargebee
  </h6>
  <p
    class="g-address__content"
  >
    <p>
      
      <span
        class="g-address_line1 g-address__value"
      >
        3rd Floor
      </span>
      <span
        class="g-address_line2 g-address__value"
      >
        Brigade Vantage 1
      </span>
      <span
        class="g-address_line3 g-address__value"
      >
        OMR Service Rd
      </span>
      
    </p>
    <p>
      
      <span
        class="g-address_city g-address__value"
      >
        Chennai
      </span>
      <span
        class="g-address_zip g-address__value"
      >
        600028
      </span>
      
    </p>
    <p>
      
      <span
        class="g-address_state g-address__value"
      >
        Tamil Nadu
      </span>
      <span
        class="g-address_country g-address__value"
      >
        India
      </span>
      
    </p>
  </p>
</div>
`;
